infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentSupervisor.java

```diff
@@ -0,0 +1,181 @@
+        long segmentId = address / StreamLogFiles.RECORDS_PER_LOG_FILE;
+
+        Path filePath = logDir.resolve(segmentId + ".log");
+
+        SegmentHandle handle = channels.computeIfAbsent(filePath, missingFilePath -> {
+            FileChannel writeCh = null;
+            FileChannel readCh = null;
+
+            try {
+                writeCh = getChannel(missingFilePath, false);
+                readCh = getChannel(missingFilePath, true);
+
+                SegmentHandle segment = new SegmentHandle(segmentId, writeCh, readCh, missingFilePath);
+                // The first time we open a file we should read to the end, to load the
+                // map of entries we already have.
+                // Once the segment address space is loaded, it should be ready to accept writes.
+                addressSpaceLoader.load(segment);
+                return segment;
+            } catch (IOException e) {
+                log.error("Error opening file {}", missingFilePath, e);
+                IOUtils.closeQuietly(writeCh);
+                IOUtils.closeQuietly(readCh);
+                throw new IllegalStateException(e);
+            } catch (RuntimeException ex) {
+                //Prevents file resources leaks in case of any RuntimeException.
+                IOUtils.closeQuietly(writeCh);
+                IOUtils.closeQuietly(readCh);
+                throw ex;
+            }
+        });
+
+        handle.retain();
+        return handle;
+    }
+
+    @Nullable
+    private FileChannel getChannel(Path filePath, boolean readOnly) throws IOException {
+        if (readOnly) {
+            if (!filePath.toFile().exists()) {
+                throw new FileNotFoundException(filePath.toString());
+            }
+
+            return FileChannel.open(filePath, EnumSet.of(StandardOpenOption.READ));
+        }
+
+        EnumSet<StandardOpenOption> options = EnumSet.of(
+                StandardOpenOption.READ,
+                StandardOpenOption.WRITE,
+                StandardOpenOption.CREATE
+        );
+        FileChannel channel = FileChannel.open(filePath, options);
+
+        // First time creating this segment file, need to sync the parent directory
+        File segFile = filePath.toFile();
+        syncDirectory(segFile.getParent());
+        return channel;
+    }
+
+    /**
+     * Close a segment
+     *
+     * @param segment segment handler
+     */
+    public void close(SegmentHandle segment) {
+        if (segment.getRefCount() != 0) {
+            log.warn("closeSegmentHandlers: Segment {} is trimmed, but refCount is {}, attempting to trim anyways",
+                    segment.getSegment(), segment.getRefCount()
+            );
+        }
+
+        channelsToSync.remove(segment.getWriteChannel());
+        channels.remove(segment.getFileName(), segment);
+        segment.close();
+    }
+
+    void sync(boolean force) throws IOException {
```

Comment:
Can you try to force all the segments instead of short-circuiting and exiting on the first failure?

Reply:
I'm not sure what could be the strategy here. If `sync` throws an exception we stop BatchProcessor.
This is how it works right now.
I'm not sure what we expect if sync fails

Reply:
If segment 1, 2 and 3 need to be synced and syncing 3 fails, that should prevent us from syncing 1 and 2. The idea is to sync as much data regardless if we will fail later on, or not. We should definitely fail, but I don't think we should fail right away.

Reply:
I totally agree but didn't want to change it because of uncertainty.
Let me fix that if we ok with the change

Reply:
done